### PR TITLE
[pipelineX](fix) Fix query cancel timeout

### PR DESCRIPTION
**Pull request by Gabriel39**
## Proposed changes

There are 2 potential reasons to cancel pipelineX query timeout.
1. Cancel fragment context first and set ready to execute will set cancel flag to false.
2. Dead lock.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Gabriel39** [comment]
run buildall

**Comment by Gabriel39** [comment]
run buildall

File: be/src/runtime/fragment_mgr.cpp [hunk 1]
```diff
@@ -574,27 +574,29 @@ Status FragmentMgr::start_query_execution(const PExecPlanFragmentStartRequest* r
...
-        LOG_INFO("Removing query {} instance {}, all done? {}", print_id(query_id),
-                 print_id(ins_id), all_done);
-        _pipeline_map.erase(ins_id);
-        g_pipeline_fragment_instances_count << -1;
     }
-    if (all_done) {
-        LOG_INFO("Query {} finished", print_id(query_id));
-        _query_ctx_map.erase(query_id);
+    {
+        std::lock_guard<std::mutex> plock(q_context->pipeline_lock);
```
Comment (yiguolei) [inline be/src/runtime/fragment_mgr.cpp:596]: 这里，拿着context 的lock，然后访问context 内部的map，感觉把这堆代码，在query context 中单独封装一个函数做更合理。

File: be/src/vec/exec/scan/scanner_context.cpp [hunk 2]
```diff
@@ -428,6 +428,7 @@ bool ScannerContext::set_status_on_error(const Status& status, bool need_lock) {
         _status_error = true;
         _blocks_queue_added_cv.notify_one();
         _should_stop = true;
+        _state->get_query_ctx()->set_exec_status(_process_status);
```
Comment (yiguolei) [inline be/src/vec/exec/scan/scanner_context.cpp:431]: why need set this？

**Review by yiguolei — APPROVED** [review]

**Review by zhiqiang-hhhh — APPROVED** [review]
LGTM

**Pull request merged**